Google.Api.CommonProtos/Rpc/StatusPartial.cs

```diff
@@ -0,0 +1,132 @@
+﻿/*
+ * Copyright 2023 Google Inc. All Rights Reserved.
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+using Google.Protobuf;
+using Google.Protobuf.Reflection;
+using Google.Protobuf.WellKnownTypes;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Google.Rpc
+{
+    public partial class Status
+    {
+        /// <summary>
+        /// Cache the full names of the messages types
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        private static class MessageNameCache<T> where T : class, IMessage<T>, new()
+        {
+            public static readonly string FullName = new T().Descriptor.FullName;
+        }
+
+        /// <summary>
+        /// Retrieves the error details of type <typeparamref name="T"/> from the <see cref="Google.Rpc.Status"/>
+        /// message.
+        /// </summary>
+        /// <remarks>
+        /// <example>
+        /// For example, to retrieve any <see cref="Google.Rpc.ErrorInfo"/> that might be in the status details:
+        /// <code>
+        ///   var errorInfo = status.GetStatusDetail&lt;ErrorInfo&gt;();
+        ///   if (errorInfo is not null) {
+        ///      // ...
+        ///   }
+        /// </code>
+        /// </example>
+        /// </remarks>
+        /// <typeparam name="T">The message type to decode from within the error details.</typeparam>
+        /// <returns>The first error details of type <typeparamref name="T"/> found, or null if not present</returns>
+        public T GetDetail<T>() where T : class, IMessage<T>, new()
+        {
+            var expectedName = MessageNameCache<T>.FullName;
+            var any = Details.FirstOrDefault(a => Any.GetTypeName(a.TypeUrl) == expectedName);
+            if (any is null)
+            {
+                return null;
+            }
+            return any.Unpack<T>();
```

Comment:
Just `return any?.Unpack<T>();` and avoid the check above? (Indeed, could do this without the `any` variable at all.)